lib/read.js

```diff
@@ -1,113 +1,87 @@
   if (def) {
     if (silent) {
       prompt += '(<default hidden>) '
-    } else if (opts.edit) {
+    } else if (edit) {
       editDef = true
     } else {
       prompt += '(' + def + ') '
     }
   }
-  var terminal = !!(opts.terminal || output.isTTY)
 
-  var m = new Mute({ replace: opts.replace, prompt: prompt })
-  m.pipe(output, {end: false})
+  const m = new Mute({ replace, prompt })
+  m.pipe(output, { end: false })
   output = m
-  var rlOpts = { input: input, output: output, terminal: terminal }
-
-  if (process.version.match(/^v0\.6/)) {
-    var rl = readline.createInterface(rlOpts.input, rlOpts.output)
-  } else {
-    var rl = readline.createInterface(rlOpts)
-  }
-
-
-  output.unmute()
-  rl.setPrompt(prompt)
-  rl.prompt()
-  if (silent) {
-    output.mute()
-  } else if (editDef) {
-    rl.line = def
-    rl.cursor = def.length
-    rl._refreshLine()
-  }
 
-  var called = false
-  rl.on('line', onLine)
-  rl.on('error', onError)
+  return new Promise((resolve, reject) => {
+    const rl = readline.createInterface({ input, output, terminal })
+    const timer = timeout && setTimeout(() => onError(new Error('timed out')), timeout)
 
-  rl.on('SIGINT', function () {
-    rl.close()
-    onError(new Error('canceled'))
-  })
+    output.unmute()
+    rl.setPrompt(prompt)
+    rl.prompt()
 
-  var timer
-  if (timeout) {
-    timer = setTimeout(function () {
-      onError(new Error('timed out'))
-    }, timeout)
-  }
-
-  function done () {
-    called = true
-    rl.close()
-
-    if (process.version.match(/^v0\.6/)) {
-      rl.input.removeAllListeners('data')
-      rl.input.removeAllListeners('keypress')
-      rl.input.pause()
+    if (silent) {
+      output.mute()
+    } else if (editDef) {
+      rl.line = def
+      rl.cursor = def.length
+      rl._refreshLine()
     }
 
-    clearTimeout(timer)
-    output.mute()
-    output.end()
-  }
-
-  function onError (er) {
-    if (called) return
-    done()
-    return cb(er)
-  }
-
-  function onLine (line) {
-    if (called) return
-    if (silent && terminal) {
-      output.unmute()
-      output.write('\r\n')
+    const done = () => {
+      rl.close()
+      clearTimeout(timer)
+      output.mute()
+      output.end()
     }
-    done()
-    // truncate the \n at the end.
-    line = line.replace(/\r?\n$/, '')
-    var isDefault = !!(editDef && line === def)
-    if (def && !line) {
-      isDefault = true
-      line = def
+
+    const onError = (er) => {
+      done()
+      reject(er)
     }
-    cb(null, line, isDefault)
-  }
+
+    rl.on('error', onError)
+    rl.on('line', (line) => {
+      if (silent && terminal) {
+        output.unmute()
+        output.write('\r\n')
+      }
+      done()
+      // truncate the \n at the end.
+      line = line.replace(/\r?\n$/, '')
+      let isDefault = !!(editDef && line === def)
+      if (def && !line) {
+        isDefault = true
+        line = def
+      }
+      return resolve([line, isDefault])
```

Comment:
the callback interface was called with `cb(err, res, isDefault)`. this is weird to do in a Promise interface since my hunch is that vast majority of usage is only concerned with `res`. meaning if we resolve the promise with an array of both properties, it forces consumers to always unwrap the array:

```js
const [res] = await read(options)
```

Since this is a breaking PR already, should we drop the `isDefault` portion of the response?

Reply:
The `isDefault` portion of the response has been removed, and the `Promise` from `read` now resolves to a string.